lib/dockerregistry/checks.go

```diff
@@ -0,0 +1,159 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package inventory
+
+import (
+	"fmt"
+	"net/url"
+	"os"
+	"strings"
+
+	gogit "gopkg.in/src-d/go-git.v4"
+	"gopkg.in/src-d/go-git.v4/plumbing"
+)
+
+// ImageRemovalCheck implements the PreCheck interface and checks against
+// pull requests that attempt to remove any images from the promoter manifests.
+type ImageRemovalCheck struct {
+	CurrentGitRepo url.URL
+	MasterSHA      plumbing.Hash
+	PullRequestSHA plumbing.Hash
+}
+
+// MKRealImageRemovalCheck returns an instance of ImageRemovalCheck.
+func MKRealImageRemovalCheck(
+	currentGitRepoStr string,
+) (*ImageRemovalCheck, error) {
+	currentGitRepo, err := url.Parse(currentGitRepoStr)
+	if err != nil {
+		return nil, fmt.Errorf("Could not parse the git repo url: %v", err)
+	}
+	masterSHAStr := os.Getenv("PULL_BASE_SHA")
+	if masterSHAStr == "" {
+		return nil, fmt.Errorf("The PULL_BASE_SHA environment variable " +
+			"doesn't exist")
+	}
+	pullRequestSHAStr := os.Getenv("PULL_PULL_SHA")
+	if pullRequestSHAStr == "" {
+		return nil, fmt.Errorf("The PULL_PULL_SHA environment variable " +
+			"doesn't exist")
+	}
+	masterSHA := plumbing.NewHash(masterSHAStr)
+	pullRequestSHA := plumbing.NewHash(pullRequestSHAStr)
+	return &ImageRemovalCheck{
+		*currentGitRepo,
+		masterSHA,
+		pullRequestSHA,
+	}, nil
+}
+
+// Run executes ImageRemovalCheck on a set of promotion edges.
+// Returns an error if the pull request removes images from the
+// promoter manifests.
+func (check *ImageRemovalCheck) Run(
+	edges map[PromotionEdge]interface{},
+) error {
+	var err error
+	// Get the Git repo the current working directory
```

Comment:
It would be great to summarize all of the Git operations and what you are doing in a short summary at the top of this area; the interspersed comments are fine, but it's hard to see why this function is doing the things it's doing. That is, focus on the forest, not so much the trees...